src/Microsoft.Health.Fhir.Core/Exceptions/CmkAadClientCredentialsGrantFailureException.cs

```diff
@@ -0,0 +1,20 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Core.Exceptions
+{
+    public class CmkAadClientCredentialsGrantFailureException : FhirException
```

Comment:
I don't think these exceptions should be in Fhir.Core since they are specific to CosmosDb

Reply:
SQL also supports Customer Managed Keys, @shmartel do you know if these would be reused if we also support this in future?

Reply:
This is a good observation.  As they were originally implemented in my first commit, I agree they were specific to Cosmos DB, as SQL would not have the same set of errors, or their errors messages would be different.  

In my 2nd commit, I now have a generic CusotmerManagedKeyException that takes an error code input, and populates the error OperationOutcomeIssue with the corresponding message.  So, at first glance, it seems that this exception could get reused for SQL, as their error codes seem to be different: https://docs.microsoft.com/en-us/sql/relational-databases/security/encryption/sql-server-connector-maintenance-troubleshooting?view=sql-server-2017#AppendixC.  So, in theory, this exception logic could check both CosmosDB and SQL error codes and return the correct error message.

But, what happens if some day the Cosmos DB and SQL error codes overlap?  Or, one of the two have some undocumented error codes that overlap.  Then the code doesn't know which error message to use.  Unlikely, but possible.

So, I will play it safe and move this to the Cosmos DB project.  I'm definitely open to suggestions on this!

Reply:
Commit 3 shows what the code looks like when CustomerManagedKeyException is moved into Microsoft.Health.Fhir.CosmosDb.  Thoughts, @namadabu , @brendankowitz ?  Given that it no longer has any Cosmos-specific types in it, I can see any argument that it could live in Core, with the caveat of an error code value collision (unlikely, but possible).

Reply:
@shmartel Now that you have made the exception generic in nature, I feel that it would be better to move this to Core.Exceptions. This way it will be consistent across both the data layers (when we support BYOK for SQL). If there is some kind of collision of status code values in the future, we can always re-factor to move the status code -> error message mapping logic to the respective data layer and still keep `CustomerManagedKeyException`. Thoughts?

Also, one issue with the current implementation is that the Api layer now needs to 'know' about the `CosmosDb` layer in order to understand `CustomerManagedKeyException`. We should avoid that kind of dependency.

Reply:
Makes sense.  I will revert my last commit which puts the exception back in Core.